mlops blog post for deploying a model to sagemaker (#3351)
pulumi/pulumi-hugo: themes/default/content/blog/mlops-huggingface-llm-aws-sagemaker-python/index.md

---
date: 2023-09-08

> title: "Deploy a Huggingface Model on AWS SageMaker with Pulumi Python IaC"

Reviewer: Is it `Huggingface` (one word) or `Hugging Face` (two words)? I've seen it both ways. The template, for example, has it as two words.

Author: I looked it up when scott mentioned it during template review. The corp is `Hugging Face Inc.` however they are also routinely marketed as either the emoji or `Huggingface`. Given the corporate name I'm okay with committing to Hugging Face and we can make huggingface all together the meta tag.

Reviewer: Fixed throughout in 46fb9a7

> title: "Deploy a Huggingface Model on AWS SageMaker with Pulumi Python IaC"

Reviewer: `s/AWS Sagemaker/Amazon Sagemaker/g`

Reviewer: Fixed via 46fb9a7

> title: "Deploy a Huggingface Model on AWS SageMaker with Pulumi Python IaC"

Reviewer 2: Hugging Face is actually two words (throughout)

allow_long_title: true

> meta_desc: "Guided short tutorial on starting a Pulumi Infrastructure as Code project to deploy Huggingface LLM Models on Amazon Web Services SageMaker Machine Learning Platform"

Reviewer: ```suggestion
meta_desc: "Guided short tutorial on starting a Pulumi infrastructure as code project to deploy Huggingface LLM Models on Amazon SageMaker machine learning platform"
```

meta_image: "meta.png"
authors:
    - kat-morgan
tags:
    - ai
    - ml
    - iac
    - aws
    - llm
    - vllm
    - mlops
    - llama
    - llama2
    - python
    - sagemaker
    - huggingface
---

> Welcome to another post in the Pulumi + Python MLOps #MLOpsChallenge series!

Reviewer 3: I really don't like the hashtag. Its not meant to be a challenge. In the previous post I called it "AI Infrastructure as Python". We can turn that into a hashtag if we want.

Reviewer 3: Also instead of focusing on the fact its a series, I would explain the concept or the problem we are solving. "Deploying a HF model on Sagemaker takes a bit of setup to spin up. Engineers run into these XYZ problems, here is how Pulumi solves it for you"

This way, you have a nice intro to why they should bother to try this out

Reviewer 4: +1. Can we move away from the MLops challenge branding, which then spills into social posts and so one? The whole concept gets messy.

We are trying to accelerate engineers spinning up AI infra. It's pretty simple.

Also, i am very keen on the post explaining the concept and problem you are solving with this: It's challenging to do X, because of this and this. Here is a magical way to solve it. 
Then you show the solution and outcome

Reviewer: I've taken the liberty of adding a quick draft on why this is important and why readers should care. @elistaykova & @aaronkao, please review and provide feedback.

Author: Can we move this type of conversation back to the campaign planning doc and associated meetings? Feedback extending beyond a single blog post content review is best facilitated on the planning end of content creation.

Author: proceeding with hashtag #PulumiMLOps per zoom call decision

> In this short tutorial we will deploy a publicly available Huggingface model on Amazon AWS SageMaker, and test a prompt using a short python script.

Reviewer: ```suggestion
In this short tutorial we will deploy a publicly available Huggingface model on Amazon SageMaker, and test a prompt using a short Python script.
```

> In this short tutorial we will deploy a publicly available Huggingface model on Amazon AWS SageMaker, and test a prompt using a short python script.

Reviewer 3: I believe their official title is Amazon Sagemaker, but pick one

Reviewer: Fixed throughout in 46fb9a7

> In this short tutorial we will deploy a publicly available Huggingface model on Amazon AWS SageMaker, and test a prompt using a short python script.

Reviewer 3: ```suggestion
In this short tutorial we will deploy a publicly available Huggingface model on Amazon AWS SageMaker, and we will test a prompt using a short python script.
```

We will be deploying:

* The [NousResearch/Llama-2-7b-chat-hf](https://huggingface.co/NousResearch/Llama-2-7b-chat-hf) model on [Hugging Face](https://huggingface.co/) 
* [Amazon AWS IAM Roles](https://www.pulumi.com/registry/packages/aws/api-docs/iam/role/)
* [Amazon AWS SageMaker Model Endpoint](https://www.pulumi.com/registry/packages/aws/api-docs/sagemaker/model/)
* [Amazon AWS CloudWatch MetricAlarm](https://www.pulumi.com/registry/packages/aws/api-docs/cloudwatch/metricalarm/)

> [Pulumi templates](../how-to-create-and-share-a-pulumi-template) enhance Infrastructure as Code (IaC) development by providing a pre-written code framework designed to help you bootstrap new pulumi projects quickly.

Reviewer: ```suggestion
[Pulumi templates](../how-to-create-and-share-a-pulumi-template) enhance the infrastructure as code (IaC) development experience by providing a pre-written code framework designed to help you bootstrap new Pulumi projects quickly.
```

> You can use the `sagemaker-aws-python` template as a working python starting point for your own AWS SageMaker deployments, and customize the model you use, SageMaker configuration, and CloudWatch integration just to scratch the surface. Your only limits are your imagination.

Reviewer: I'll stop pointing out AWS SageMaker from here on, but please make sure you catch all instances.

Reviewer: Fixed via 46fb9a7

> ## Requirements

Reviewer: Before getting into the nitty-gritty details, can we add some high-level text about why this is important? You know, to give someone some context about the business value of deploying an HF LLM to SageMaker? That'll help ground this and provide some justification for _why_ someone should care about this post.

Reviewer 3: Agreed, its about what is the challenge and difficulty of doing this the alternative/normal way. Then why Pulumi provides an easier solution

* [Python3 (3.9+)](https://www.python.org/downloads/)
* [Pulumi CLI](https://www.pulumi.com/docs/install/)
* [Pulumi account](https://app.pulumi.com/signup)
* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)
* [Configured & Logged in with AWS Credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-authentication.html)

## Instructions

### 1. Login to Pulumi Cloud and initialize stack

Let's begin by logging into Pulumi Cloud

```bash
# There are many ways to store Pulumi state, here we use Pulumi Cloud
# Other state backends include S3, local file, and more
pulumi login
```

> ### 2. Prepare a new Pulumi Project

Reviewer 3: isn't there a `pulumi config set aws:region <REGION>` somewhere in here. Yours goes direct to us-east-1, but I think you have to set it first right?

Author: it's a default provided by the template, user's are prompted to accept the default or override with their desired region when creating the template.

> Here we create our new pulumi project directory and populate it from the `sagemaker-aws-python` [Pulumi template](https://github.com/pulumi/templates)

Reviewer: ```suggestion
Here we create our new Pulumi project directory and populate it from the `sagemaker-aws-python` [Pulumi template](https://github.com/pulumi/templates)
```

```bash
# Create a new directory & change directories into it
mkdir newSagemaker && cd newSagemaker

# Start your project from the sagemaker-aws-python template
# Follow along with the prompts to create your new project and initialize a stack
pulumi new sagemaker-aws-python
```

### 3. Deploy your model as a new SageMaker endpoint

> This step may take between 10 and 20 minutes while Amazon AWS builds your infrastructure. You can follow along in the console as resources are provisioned, or open the link displayed in terminal for your stack in a browser to follow along there too!

Reviewer: ```suggestion
This step may take between 10 and 20 minutes while Amazon builds your infrastructure. You can follow along in the console as resources are provisioned, or open the link displayed in terminal for your stack in a browser to follow along there too!
```

```bash
pulumi up

> ```

Reviewer 3: can we show a quick image or just text of what the pulumi up output looks like?

Reviewer 4: +1. That was my first reaction after the reading the post. So what? I think a lot (most) of the code you are showing can be in the readme. Instead, show the interesting infrastructure bits and the resources, the output, what/how was created. Instead of all the details where and how to login

Author: done

## Try your new SageMaker Endpoint

Once your stack has finished deploying, use this rudimentary Python snippet to test the deployed SageMaker endpoint.

> > NOTE: notice that we are using `us-east-1` in this script, be sure to change the reagion in python to match the reagion you deployed the SageMaker endpoint into.

Reviewer: ```suggestion
> NOTE: notice that we are using `us-east-1` in this script, be sure to change the reagion in python to match the region you deployed the SageMaker endpoint into.
```

### 4. Save the following python snippet as `test.py`

```python
import json, boto3, argparse

def main(endpoint_name):
    client = boto3.client('sagemaker-runtime', region_name='us-east-1')
    payload = json.dumps({"inputs": "In 3 words, name the biggest mountain on earth?"})
    response = client.invoke_endpoint(EndpointName=endpoint_name, ContentType="application/json", Body=payload)
    print("Response:", json.loads(response['Body'].read().decode()))

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("endpoint_name")
    main(parser.parse_args().endpoint_name)
```

### 5. Run the test prompt

> Now source the local `venv` and run the script using the AWS name of your new endpoint from pulumi state outputs:

Reviewer: ```suggestion
Now source the local `venv` and run the script using the name of your new endpoint from Pulumi state outputs:
```

```bash
# On Linux & MacOS
source venv/bin/activate

# Execute test.py
python3 test.py $(pulumi stack output EndpointName)
```

### 6. Cleanup / Teardown

> Finally, be sure to clean up un-used resources when you're done with one easy command.

Reviewer: ```suggestion
Finally, when you're finished with testing your Hugging Face model on SageMaker, you can easily clean up un-used resources with one easy command.
```

```bash

> pulumi destroy

Reviewer 3: ```suggestion
pulumi destroy -rf --remove
```

i normally do that so they can just nuke the whole thing in one command

Author: no, this approach is great in a gist or developer docs, but not when we are handholding a new-to-iac audience.

```

## Conclusion

To recap, in a few commands, we created a new Pulumi project from a ready to roll template, deployed an LLM Model endpoint on AWS SageMaker, and tested it with a short python script to generate a response from our model!

> AI and ML is becoming more of a necessity every day. It may appear daunting or out of reach at first glance, but with the power of IaC written in Pulumi Python programs, getting started has never been easier. 

Reviewer: This is the kind of high-level justification I mentioned having earlier in the post. Maybe move this up there, or copy it up there and recap it here as a reminder?

Reviewer 3: well this justification is still too nebulous. i would like to see the exact details and customer pain

[venv]:/docs/languages-sdks/python/#virtual-environments

[Python]:https://www.python.org
[Python3]:https://www.python.org/downloads
[natural language prompt]:https://en.wikipedia.org/wiki/Prompt_engineering
[Meta AI LlaMa 2]:https://ai.meta.com/llama
[Hugging Face]:https://huggingface.co
[NousResearch/Llama-2-7b-chat-hf]:https://huggingface.co/NousResearch/Llama-2-7b-chat-hf
[AWS CLI]:https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html
[AWS Credentials]:https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-authentication.html
[Amazon SageMaker]:https://aws.amazon.com/pm/SageMaker
[tremendous value]:https://blogs.nvidia.com/blog/2023/01/26/what-are-large-language-models-used-for
[Large Language Models]:https://en.wikipedia.org/wiki/Large_language_model
[LLM]:https://en.wikipedia.org/wiki/Large_language_model
Welcome to another installment of [Pulumi Python] #MLOpsChallenge!

In this short tutorial we will deploy a publicly available [Meta AI LlaMa 2] based model from [Hugging Face], on [Amazon SageMaker]. Then we will test it with a [natural language prompt] using a short [Python] script.

> It may be tempting to dismiss AI as yet another over hyped and complicated technology, but is it really? What if you could start using this cutting edge technology without the stress? AI/ML models, including [Large Language Models] (LLM) like what we will demonstrate today, offer [tremendous value] to organizations, but only if AI/ML platform resources can be reliably, easily, and cost effectively provisioned and put to work. By utilizing [Infrastructure as Code] (IaC) written as [Pulumi Python] programs, machine learning operations teams benefit from [IaC] making deploying AI applications the easiest part of your AI/ML journey. Benefits of [Pulumi Python] based [IaC] for your AI/ML projects include:

Reviewer 3: ```suggestion
Running Hugging Face models on Amazon SageMaker is a popular option for deploying AI/ML model backends. However, much of the infrastructure setup, configuration, and deployment is done manually via the SageMaker console. This leads to a number of issues such as manual errors or can your configuration be repeated easily for many deployment. Infrastructure as Code allows you to define all your infrastructure configurations as piece of code that can be version controlled and then the provisioning can be automated. With Pulumi, you can use Python which you are using for everything else to write IaC programs. 
```

> 1. Easily create AI/ML infrastructure, together with other supporting infrastructure _in the same Pulumi program_

Reviewer 3: after the above para, doesn't seem this is that necessary

2. [Pulumi templates] enhance the infrastructure as code (IaC) development experience by providing a pre-written code framework designed to help you bootstrap new Pulumi projects quickly.
3. Cloud orchestration written in programming languages like [Python] which is already well known and practiced in the AI/ML industry.

## Pulumi makes AI/ML Easy?

For this exercise we will deploy a [Pulumi Template] from scratch. Starting from the `pulumi new` command, we will initialize a new [Pulumi project and stack]. This will provide everything required to deploy our chosen Hugging Face model as an [Amazon SageMaker Model Endpoint]

> You can use the `sagemaker-aws-python` template as a working Python starting point for your own Amazon SageMaker deployments, and customize the model you use, SageMaker configuration, and CloudWatch integration as well, just to scratch the surface. Your only limits are your imagination.

Reviewer 3: combine this paragraph with the above. 

"For this exercise we will deploy the sagemaker-aws-python template as a working Python starting point.... You will use the pulumi new command to launch this template. This will launch X, Y, Z. Use it as is or as a starting point for bigger infrastructure configurations...

Now let's look at what the `sagemaker-aws-python` template will deploy in this tutorial:

> * [AWS IAM Roles]

Reviewer 3: I would do this maybe in pargraph form so there is more a relationship between everything. 

The template will deploy a Sagemaker instance. IaM roles are attached so it can do <what>. Then huggingface models are downloaded and then mapped to the instance. CW alarms are added to prevent <over cost charges???>. Then an endpoint is exposed and given to you in the outputs. The end point is serving the model
